scrapy/pipelines/files.py

```diff
@@ -248,6 +250,42 @@ def persist_file(self, path, buf, info, meta=None, headers=None):
         )
 
 
+class FTPFilesStore(object):
+    
+    def __init__(self, uri):
+        assert uri.startswith('ftp://')
+        u = urlparse(uri)
+        self.ftp = FTP()
+        self.ftp.connect(u.hostname, u.port or '21')
+        self.ftp.login(u.username, u.password)
+        self.basedir = u.path + '/'
+        ftp_makedirs_cwd(self.ftp, self.basedir+'full')
+        
+    def persist_file(self, path, buf, info, meta=None, headers=None):
+        buf.seek(0)
+        filename = path.split('/')[1]
+        return threads.deferToThread(
+            self.ftp.storbinary,
+            'STOR %s' % filename,
+            buf
+        )
+            
+    def stat_file(self, path, info):
+        def _stat_file(path):
+            try:
+                last_modified = float(self.ftp.voidcmd("MDTM " + self.basedir + '/' + path)[4:].strip())
+                m = hashlib.md5()
+                self.ftp.retrbinary('RETR %s' % self.basedir + path, m.update)
+                return {'last_modified': last_modified, 'checksum': m.hexdigest()}
+            # The file doesn't exist
+            except Exception as e :
+                return {} 
+        return threads.deferToThread(_stat_file, path)
+    
+    def close_connection(self):
```

Comment:
Is this really necessary? Isn’t it better to just let the connection remain open until the `self.ftp` object gets destroyed naturally?

Reply:
Let me check that :)

Reply:
The `FTP` class does the cleanup in its `__exit__` method, so yeah it is not necessary like you said :D